app_dart/lib/src/request_handlers/check_flaky_tests_and_update_github.dart

```diff
@@ -0,0 +1,251 @@
+            _isOtherIssueMoreImportant(nameToExistingIssue[match.namedGroup('name')], issue)) {
+          nameToExistingIssue[match.namedGroup('name')] = issue;
+        }
+      }
+    }
+    return nameToExistingIssue;
+  }
+
+  Future<Map<String, PullRequest>> _getExistingPRs(GithubService gitHub, RepositorySlug slug) async {
+    final Map<String, PullRequest> nameToExistingPRs = <String, PullRequest>{};
+    for (final PullRequest pr in await gitHub.listPullRequests(slug, null)) {
+      final RegExpMatch match = pullRequestTitleRegex.firstMatch(pr.title);
+      if (match != null) {
+        nameToExistingPRs[match.namedGroup('name')] = pr;
+      }
+    }
+    return nameToExistingPRs;
+  }
+
+  Set<String> _getImportantFlakes(List<BuilderStatistic> statisticList, double threshold) {
+    final Set<String> importantFlakes = <String>{};
+    for (final BuilderStatistic statistic in statisticList) {
+      if (statistic.flakyRate > threshold) {
+        importantFlakes.add(statistic.name);
+      }
+    }
+    if (importantFlakes.isNotEmpty) {
+      return importantFlakes;
+    }
+    // No flake is above threshold.
+    BuilderStatistic mostImportant;
+    for (final BuilderStatistic statistic in statisticList) {
+      if (mostImportant == null || mostImportant.flakyRate < statistic.flakyRate) {
+        mostImportant = statistic;
+      }
+    }
+    return <String>{
+      if (mostImportant != null) mostImportant.name,
+    };
+  }
+
+  Future<void> _updateFlakes(
+    GithubService gitHub,
+    RepositorySlug slug, {
+    @required BuilderStatistic statistic,
+    @required String owner,
+    @required Issue existingIssues,
+    @required bool hasExistingPR,
+  }) async {
+    // Don't create a new issue if there is a recent closed issue within
+    // kGracePeriodForClosedFlake days. It takes time for the flaky ratio to go
+    // down after the fix is merged.
+    Issue issue = existingIssues;
+    if (issue == null ||
+        (issue.state == 'closed' &&
+            DateTime.now().difference(issue.closedAt) > const Duration(days: kGracePeriodForClosedFlake))) {
+      final IssueBuilder issueBuilder = IssueBuilder(statistic: statistic, threshold: _threshold, isImportant: true);
+      issue = await gitHub.createIssue(
+        slug,
+        title: issueBuilder.issueTitle,
+        body: issueBuilder.issueBody,
+        labels: issueBuilder.issueLabels,
+        assignee: owner,
+      );
+    }
+    if (issue != null) {
+      if (!hasExistingPR && !await _isAlreadyMarkedFlaky(gitHub, slug, statistic.name)) {
+        final String modifiedContent =
+            _marksBuildFlakyInContent(await gitHub.getFileContent(slug, kCiYamlPath), statistic.name, issue.htmlUrl);
+        final GitReference masterRef = await gitHub.getReference(slug, kMasterRefs);
+        final PullRequestBuilder prBuilder = PullRequestBuilder(statistic: statistic, issue: issue);
+        final PullRequest pullRequest = await gitHub.createPullRequest(slug,
```

Comment:
Just recalled that for shard tests that contain thousands of unit tests, we may want to skip the step to create a PR. Test owner or the tree gardener will need to look into the flaky test and determine which sub test causes the flake, and then they need to create a PR to disable the sub test instead of marking the whole shard test suite as flaky.

From the auto process side, it will be challenging to find out the sub test and disable it based on current test framework.

Reply:
is there a way to determine which tests are shard tests?

Reply:
most ones here: https://github.com/flutter/flutter/blob/master/TESTOWNERS#L198
```
build_tests @zanderso @flutter/tool
framework_tests @HansMuller @flutter/framework
tool_integration_tests @zanderso @flutter/tool
tool_tests @zanderso @flutter/tool
web_integration_tests @ferhatb @flutter/web
web_long_running_tests @ferhatb @flutter/web
web_tests @ferhatb @flutter/web
web_tool_tests @zanderso @flutter/tool
```
Each of them contains several subshards, which are mapped to the builder name in .ci.yaml. So we will check those subshards flakiness and only file flaky issues rather than create a PR to mark them flaky. (each subshard contains tons of unit tests).